HW0/P5/P5.txt

```diff
@@ -1 +1,32 @@
-# Your answers here.
+1.  When two bags are added together, we can think of them as forming a new bag with value equal to the original two bags combined, so it's just like adding numbers. There are 255 additions required to add all 256 bags. After one second, bags 1 and 2 have been added, and there are 255 bags left: bag (1 and 2), and bags 3 through 255. After two seconds, 254 bags remain. Similarly, after 255 additions, bags 1 through 255 will have been added with bag 256, and there is only one bag left. Since one cashier must perform all 255 additions, it will take them 255 seconds.
+
+2.  There are still 255 additions to perform. Each round, a cashier can add two bags together, so each round, the eight cashiers can add a total of 16 bags. After 256/16=16 rounds, each cashier will have added 16*2=32 bags, and there will be 128 bags waiting to be added. After a further 128/16=8 rounds, there will be 64 bags left. Four rounds later, there will be 32 bags; two rounds after that, there will be 16 bags; one round later, there will be eight bags.
+
+Now the cashiers cannot all work at the same time, so it takes one round to go from eight to four, one round to go from four to two, and one round to add the remaining two bags together. Thus, it takes 16+8+4+2+1+1+1+1=34 seconds total.
+
+3.  Infinite cashiers can halve the number of bags every second, since bags can be paired off and each pair added together. It will therefore take log_2(256) = 8 seconds.
+
+4.  As above, it will take log_2(N) seconds.
+
+5.  I will refer to communicating a number as giving a bag, since in this problem, they're essentially the same thing. That is, adding the value of the pennies in two bags is no different than adding the pennies in the bags together and noting their value.
+	 After the first round, there will be 128 numbers to be added. Half of the cashiers will then add their number to another cashier's number, and a cashier will take one second to communicate their number. Thus, when there are 128 numbers to be added, there will be an extra 128/2=64 seconds of time before the next round of adding can take place. It will still take eight rounds of just adding, but now there are an extra 64+32+16+4+2+1=119 seconds of overhead. The total time required is therefore 127 seconds, about eight times as slow as the case when communication was instant.
```

Comment:
In hindsight, this is totally wrong, since cashiers can communicate in parallel.